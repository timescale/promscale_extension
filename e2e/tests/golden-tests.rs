use crate::common::{exec_sql_script, run_postgres};
use insta::assert_snapshot;
use std::env;
use test_generator::test_resources;
use testcontainers::clients;

mod common;

#[test_resources("testdata/*.sql")]
fn golden_test(resource: &str) {
    let docker = clients::Cli::default();
    let node = run_postgres(&docker);
    let query_result = exec_sql_script(&node, resource);

    assert_snapshot!(resource, query_result);
}
