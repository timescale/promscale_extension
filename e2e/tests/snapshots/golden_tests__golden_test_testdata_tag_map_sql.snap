---
source: e2e/tests/golden-tests.rs
expression: query_result
---
\set ON_ERROR_STOP 1
CREATE EXTENSION promscale;
CREATE EXTENSION
SELECT put_tag_key('char', ps_trace.span_tag_type());      /* 1001 */
 put_tag_key 
-------------
        1001
(1 row)

SELECT put_tag_key('iteration', ps_trace.span_tag_type()); /* 1002 */
 put_tag_key 
-------------
        1002
(1 row)

SELECT put_tag_key('pwlen', ps_trace.span_tag_type());     /* 1003 that is used */
 put_tag_key 
-------------
        1003
(1 row)

INSERT INTO _ps_trace.tag(id,tag_type,key_id,key,value)
    OVERRIDING SYSTEM VALUE
    VALUES
    (93,2,6,'telemetry.sdk.language','"python"'),
    (94,2,5,'telemetry.sdk.name','"opentelemetry"'),
    (95,2,7,'telemetry.sdk.version','"1.8.0"'),
    (114,2,1,'service.name','"generator"'),
    (242,5,1003,'pwlen','18'),
    (247,5,1003,'pwlen','25');
INSERT 0 6
INSERT INTO _ps_trace.span(trace_id,span_id,parent_span_id,operation_id,start_time,end_time,duration_ms,trace_state,span_tags,dropped_tags_count,event_time,dropped_events_count,dropped_link_count,status_code,status_message,instrumentation_lib_id,resource_tags,resource_dropped_tags_count,resource_schema_url_id)
    VALUES
        (E'78dd078e-8c69-e10a-d2fe-9e9f47de7728',-2771219554170079234,NULL,19,E'2022-04-26 11:44:55.185139+00',E'2022-04-26 11:44:55.38517+00',200.031,NULL,E'{"1003": 247}',0,E'["2022-04-26 11:44:55.185659+00","2022-04-26 11:44:55.385148+00")',0,0,E'error',E'Exception: FAILED to fetch a lower char',5,E'{"1": 114, "5": 94, "6": 93, "7": 95}',0,NULL),
        (E'05a8be0f-bb79-c052-223e-48608580efce',2625299614982951051,NULL,19,E'2022-04-26 11:44:55.185962+00',E'2022-04-26 11:44:55.288812+00',102.85,NULL,E'{"1003": 242}',0,E'["2022-04-26 11:44:55.185999+00","2022-04-26 11:44:55.288781+00")',0,0,E'error',E'Exception: FAILED to fetch a lower char',5,E'{"1": 114, "5": 94, "6": 93, "7": 95}',0,NULL);
INSERT 0 2
/* tag_map_denormalize function is actually called */
SELECT span_tags, resource_tags FROM ps_trace.span;
   span_tags   |                                                               resource_tags                                                                
---------------+--------------------------------------------------------------------------------------------------------------------------------------------
 {"pwlen": 25} | {"service.name": "generator", "telemetry.sdk.name": "opentelemetry", "telemetry.sdk.version": "1.8.0", "telemetry.sdk.language": "python"}
 {"pwlen": 18} | {"service.name": "generator", "telemetry.sdk.name": "opentelemetry", "telemetry.sdk.version": "1.8.0", "telemetry.sdk.language": "python"}
(2 rows)

CREATE FUNCTION explain_jsonb(_query pg_catalog.text, OUT _result pg_catalog.jsonb)
    RETURNS pg_catalog.jsonb
    LANGUAGE plpgsql VOLATILE AS
$fnc$
BEGIN
    EXECUTE 'explain (format json) ' || _query INTO _result;
    RETURN;
END;
$fnc$;
CREATE FUNCTION
/* Not equals operator returns correct result */
PREPARE neq_test AS
SELECT trace_id
    FROM ps_trace.span
    WHERE
            span_tags -> 'pwlen' != '25'::jsonb
        AND resource_tags -> 'service.name' = '"generator"';
PREPARE
EXECUTE neq_test;
               trace_id               
--------------------------------------
 05a8be0f-bb79-c052-223e-48608580efce
(1 row)

/* Not equals uses support function and produces correct plan */
SELECT 
    x::text LIKE '%InitPlan%',
    x::text LIKE '% @> ANY%' 
FROM explain_jsonb('EXECUTE neq_test') AS f(x);
 ?column? | ?column? 
----------+----------
 t        | t
(1 row)

/* Equals operator returns correct result */
PREPARE eq_test AS
SELECT trace_id
    FROM ps_trace.span
    WHERE
            span_tags -> 'pwlen' = '25'::jsonb
        AND resource_tags -> 'service.name' = '"generator"';
PREPARE
/* Equals uses support function and produces correct plan */
SELECT 
    x::text LIKE '%InitPlan%',
    x::text LIKE '% @> %' 
FROM explain_jsonb('EXECUTE eq_test') AS f(x);
 ?column? | ?column? 
----------+----------
 t        | t
(1 row)

/* Functions, underpinning tag_map operators handle NULL arguments correctly */
SELECT ps_trace.tag_v_eq(NULL, NULL);
 tag_v_eq 
----------
 
(1 row)

SELECT ps_trace.tag_v_ne(NULL, NULL);
 tag_v_ne 
----------
 
(1 row)

SELECT ps_trace.tag_map_object_field(NULL, NULL);
 tag_map_object_field 
----------------------
 
(1 row)


